### PR TITLE
Add metric notebook for probability to fail

### LINKED ISSUE #282: KPI Request : Probability to Fail
As an OpenShift manager, I would want to know for a given moment in time, what is the probability that the test will fail. Using this information, we plan to build a model that uses the probability to fail to predict the time to failure. (#226 )

Acceptance Criteria : 
- [ ] small notebook in `notebooks/data-sources/TestGrid/metrics/` that collects this metric and stores it as a csv
- [ ] The metric should store `probability-to-fail`.

Approach : 
For a test (or set of tests), look at the distribution of test duration for all failed tests and the distribution of test duration for all the successful tests (ideally generating two distinct distributions). Then given a new duration value, evaluate the likelihood that it belongs to one of the two distributions. This will give a way to calculate the "probability to fail" based solely on the duration metric.

### PR DESCRIPTION
**Pull request by aakankshaduggal**
This PR adds a metric notebook that calculates `probability_to_fail`.

Co-authored-by: Aakanksha Duggal <aduggal@redhat.com>
Co-authored-by: Hema Veeradhi <hveeradh@redhat.com>

## Related Issues and Dependencies

Fixes #282 

## This introduces a breaking change

- [ ] Yes
- [x] No

### CODE REVIEW

File: notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb [hunk 1]
```diff
@@ -0,0 +1,1091 @@
+{
```
Comment (oindrillac) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Can you re-iterate the problem statement here? as in what are you trying to calculate and why did we choose Calibrated Classification model to be a suitable approach here as opposed to any other standard classification technique? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (oindrillac) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Should we sample df after applying p_f_label and removing -1s or NaNs so we get 1000 meaningful records </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (oindrillac) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>related to the above comment: but what does predicting a -1 label mean here? should we rename <code>pass_or_fail</code> then to something like <code>pass_fail_or_neither</code> depending on what it signifies as its a bit misleading.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (oindrillac) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>If this is multi-class classification with 3 classes, will negating the probabilities give us accurate results?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (oindrillac) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>&gt;&gt;&gt;&gt; suggested change</p><p> </p><p>In this notebook, we calculated for a given moment in time, what is the probability that a test will fail. Using this information, we plan to build a model that uses the probability to fail value to predict the time to failure, that is how many days, weeks, months, hours, minutes or seconds do you have until the test in question stops working. As next steps, we plan to work on this <a href="https://github.com/aicoe-aiops/ocp-ci-analysis/issues/226" target="_blank">issue</a>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>We are just assigned <code>-1</code>  to the NaN records. Because the model needs a numeric value. </p>

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Yes, we should remove all <code>-1</code>  as they aren't really usable since they represent all other outcome types as well as still running tests, right?</p>

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>+1</p>

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Why don't you simply remove the NaN records? They represent a number of different label types and might make learning a probability of pass or fail more difficult. </p>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Sure. We will add the explanation.</p>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Sure. I agree. Just that we were losing many values like that when we take the entire dataset. But yes, I will remove the NaN records instead.</p>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Sure. We will remove that</p>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>I think if we remove the NaN, we will have only two classes. This should solve this problem.</p>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Will add this. Thank you so much for your suggestions :)</p>

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p><code>probs_test = calibrated.predict_proba(testx)[:, 1]</code>  </p><p> </p><p>I assume that this is outputting an array with 3 values, on for each class, why not just collect <code>[:,0]</code>   which is probably the failure class from the onset? </p>

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Also, are these true probabilities? Like does <code>sum(calibrated.predict_proba(testx)[:, 1])</code>  always equal 1.0?</p>

Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.    AUTOMATION = os.getenv("True")
</pre><p>Shouldn't this be <code>AUTOMATION = os.getenv("IN_AUTOMATION")</code>  <span class="ql-cursor">﻿</span></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>What is the reason for using the CalibratedClassifierCV in addition to the SVM? You can still get probabilities from the svm using <code>model.predict_proba()</code></p><p> </p><p>It just seems like an extra step where its unclear its needed/ provides more accurate results than using svm directly.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.        X, y, train_size=0.8, test_size=0.2, random_state=None
</pre><p>Why not provide a random state? Without it your results will be more difficult to reproduce.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #20.    probs_test = calibrated.predict_proba(testx)[:, 1]
</pre><p>What is the testing accuracy/precision/recall/f1 for this method? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Do we need to retain <code>test_duration</code>  and <code>pass_or_fail</code>  here? Aren't they stored in other metric files? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>That's a great suggestion. We are on it :)</p>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: The `test_duration` we are using is different from all other files and `pass_or_fail` also we have added for the first time.

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: We have not calculated it, but we can add that.

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: I agree, we can add that too.

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: :+1:

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>The dataset was originally around 21M values and now we have only 8M :/</p>

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: 8M should still be more then enough for this use case

Comment (hemajv) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>yes, we will add the graphs which explain the difference between how the probabilities are calculated with and without calibration</p>

Comment (hemajv) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>so the <code>pass_or_fail</code>  column was added to convert the boolean values to 1 or 0 for passing and failing tests respectively so that they can be used to calculate the probabilities. In the <code>test_pass_fialures.ipynb</code>  notebook, we are just retaining the boolean values as-is</p>

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: Sure, but the translation from True/False to 1/0 is trivial. So I don't see why we need to store both a boolean and numerical representation of the same information in Ceph. Once you've calculated the metric of interest (probability to fail), what is the value of retaining columns with the interim values?

Comment (hemajv) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>ah yes my bad, we can drop this column before saving it to ceph</p>

File: notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb [hunk 2]
```diff
@@ -0,0 +1,1191 @@
...
```
Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Just curious, why don't you normalize the probabilities here?</p><p> </p><p>Also doesn't <code>probability = True</code> normalize them?</p><pre class="ql-syntax" spellcheck="false">Line #14.    model = SVC(probability=True).fit(trainx, trainy)
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Does the Brier score work correctly if the probabilities are not normalized? Looks like it is the mean squared difference of the probabilities from their target (either 0.0 or 1.0) So if the output, ypred, is not a probability, then the results of this score will be misleading, right?</p><p> </p><p>https://github.com/scikit-learn/scikit-learn/blob/2beed5584/sklearn/metrics/_classification.py#L2510</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: <p>Do you think you could improve you results, for accuracy , briair score, roc, and the reliability diagram if you resampled your training data to be balanced? Currently it is 96.7% pass and 3.2% fail. So according to the accuracy score in both cases this is modeling the distribution of the labels and not exhibiting much predictive power.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/310/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: No, it actually enables the probability estimates(from the methods `predict_proba` and `predict_log_proba`) because SVMs do not directly provide probability estimates.

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: Yes, I totally agree that these metrics are not giving us the results we are looking for because these are just metrics for the correct classification of the labels `pass` or `fail` and not the probability of getting classified into the classes.

In order to get a metric for the probability model, we have created the calibration curve that allows us to choose a better model in this scenario.

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: Yes, I can try that out and see if that gives us better results. 👍

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: I'm a little confused. What is the difference between a "normalized probability" and "probability estimates" ?

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: :+1:

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: Probability estimates here are the calibrated probabilities from SVM and they are different from normalized probabilities in the sense that their summation is not equal to 1. The output of the predict_proba method can be directly interpreted as a confidence level, i.e. among the samples to which it gave a predict_proba value close to 0.8, approximately 80% actually belong to the positive class.

Comment (aakankshaduggal) [reply notebooks/data-sources/TestGrid/metrics/probability_to_fail.ipynb:1]: We tried various resampling techniques and by undersampling, we achieved better results for the three metrics we were calculating, so we will add an updated notebook with resampled data. Thanks for the suggestion @MichaelClifford

**Comment by MichaelClifford** [comment]
:+1:  Thanks for making the updates :) 

/approve
